### PR TITLE
Update invenio GitHub

### PR DESCRIPTION
**Pull request by alejandromumo**
closes https://github.com/inveniosoftware/invenio-github/issues/98

This PR is based on the updates from https://github.com/inveniosoftware/invenio-github/pull/97

### CODE REVIEW

File: invenio_github/config.py
Comment (alejandromumo) [inline invenio_github/config.py:1]: I did not spend a lot of time looking at these configs. 
I think some can be removed and others are not being picked up (e.g. `GITHUB_REMOTE_APP["signup_handler"]["setup"]`)

File: invenio_github/api.py
Comment (alejandromumo) [inline invenio_github/api.py:384]: These three class methods were not tested yet by me. 

It seems only `revoke_token` is used by a task

File: invenio_github/receivers.py
Comment (alejandromumo) [inline invenio_github/receivers.py:150]: This was encountered before implementing a "state" in the receiver. This "state" act as a lock for a release. E.g. if an event already was received for that release, we do not process the event anymore.

So probably the catch for this generic `SQLAlchemyError` can be removed, however, has to be tested.

I also checked legacy Zenodo and saw some events that failed due to this error.

File: invenio_github/tasks.py
Comment (alejandromumo) [inline invenio_github/tasks.py:74]: I am not sure what to do with this task, it should probably be a post processor on the disconnect signal emitted by invenio-oauthclient

File: invenio_github/ext.py
Comment (alejandromumo) [inline invenio_github/ext.py:80]: question: do we want these behaviours? 

What happens is the following:

- User connects Github account (invenio-oauthclient)
- A signal is emitted and we start syncing user's repositories, initialising their account.

The alternative is not to sync anything automatically, instead, it's synced on demand (e.g. an HTTP request)

File: invenio_github/views/github.py
Comment (alejandromumo) [inline invenio_github/views/github.py:51]: question: this decorator instantiates a `GithubAPI` object, while the views that use the decorator start by instantiating it (again). Should we pass the instance? E.g.

```python
kwargs["github_api"] = github
return f(*args, **kwargs)
```

File: invenio_github/views/github.py
Comment (alejandromumo) [inline invenio_github/views/github.py:77]: TODO: the github icon does not show up in the UI. This has to be enabled when working the UI part of the integration

File: invenio_github/views/github.py
Comment (alejandromumo) [inline invenio_github/views/github.py:135]: I'm not a big fan of these `try/catch` and then `abort(code)`. Perhaps we can have error handlers (as we do with resources). For now it works, but there's room for improvement

File: invenio_github/views/github.py
Comment (alejandromumo) [inline invenio_github/views/github.py:141]: Do we want to separate `ui` and `api` endpoints? We do this separation in some modules, it's easier to navigate the code.

File: invenio_github/views/github.py
Comment (alejandromumo) [inline invenio_github/views/github.py:163]: I moved away from nested sessions, as they were giving me some trouble. Namely, when we have to commit a session before spawning a celery task it would make the nested session error on exit.

File: run-tests.sh
Comment (alejandromumo) [inline run-tests.sh:27]: We can copy/paste our `run-tests.sh` from other modules

File: setup.cfg
Comment (alejandromumo) [inline setup.cfg:63]: `invenio-webhooks` needs to be released (major, `2.0.0`) and we need to bump it here.

File: setup.cfg
Comment (alejandromumo) [inline setup.cfg:62]: It seems that `invenio_rest` is not released, it has some changes in master.

However I am not sure we need this dependency, we're just import `RESTException` and using it once.

File: setup.cfg
Comment (alejandromumo) [inline setup.cfg:45]: many dependencies might not be needed, I just bumped what we had and removed the most obvious ones. 

e.g. `mistune` is not used, etc. I can have a look afterwards, for not it's not blocking

File: tests/test_badge.py
Comment (alejandromumo) [inline tests/test_badge.py:29]: Commented for now since I did not touch badges and tests were failing

File: invenio_github/errors.py
Comment (alejandromumo) [inline invenio_github/errors.py:63]: I think this is old way of subclassing in python. 


```suggestion
        super().__init__(message or self.message)
```

applies for all this file

File: invenio_github/errors.py
Comment (ppanero) [inline invenio_github/errors.py:35]: This string does not include any of the attributes of the error class. Are those included somehow by the parent? (Happens also to other errors in this file)

File: invenio_github/ext.py
Comment (ppanero) [inline invenio_github/ext.py:80]: The fact that we have to access an underscored method somehow tells that we should do it differently?

File: invenio_github/models.py
Comment (ppanero) [inline invenio_github/models.py:184]: what about `one_or_none` so higher level classes dont need to deal with SQL (data layer) exceptions like `NoResultsFound`?

File: invenio_github/receivers.py
Comment (ppanero) [inline invenio_github/receivers.py:88]: This seems to have the full logic of the outer function, why do we need a function inside a function.

File: invenio_github/receivers.py
Comment (ppanero) [reply invenio_github/receivers.py:150]: I agree on removing it, at the moment it serves as catch all for errors we might not be aware about.

File: invenio_github/tasks.py
Comment (ppanero) [inline invenio_github/tasks.py:80]: this points to a clear design/separation issue 😅 😇

File: invenio_github/utils.py
Comment (ppanero) [inline invenio_github/utils.py:1]: It looks like this utils functions can be modeled as api classes (or resolvers 🤔 ), in most of the cases we need a wrapper over a GH entity or to actually get it (resolve it). And we could 🔥 🧨 the `utils.py`.

File: invenio_github/utils.py
Comment (ppanero) [inline invenio_github/utils.py:86]: why is there a 30 limit? should it be configurable?

File: invenio_github/views/github.py
Comment (ppanero) [inline invenio_github/views/github.py:1]: remove all imports with `from __future__`

File: run-tests.sh
Comment (ppanero) [reply run-tests.sh:27]: I agree, to be consistent

**Review by ppanero — COMMENTED** [review]
I've reviewed this with the premise that is a module migration to work with RDM but is not the intention to leave the code 100% clean and fix previous problems (although many were done anyway). I tried to review as much as possible, but is very hard to make an in-depth review in a sensible time due to its size; so my comments are more on structure than actual code implementation.

Tiny rant, nothing personal:

I'm seeing this (very big PRs, I've done it too), and I think we are setting ourselves for failure. Big PRs stay very long in the review column since they are very intimidating. For future reference, it would be nice to have more self contained PRs.
- Black and other migrations
- Example app removal
- Actual code/functionality changes

Having all in one PR with 54 files and 2.2/2.7k changes is quite unmanagable and very time consuming to review. Half of the brain power goes to understand if the change was black formatting or actual code. We can review commits, but in this case there's 5 on the actual code so there's still a need to jump from commit to commit.

Honestly, even GitHub lags to handle a PR of this size 😂 

Comments apply in many places related to layers:

- We are importing `NoResultsFound` in some classes and in some we are not.
- There is an `api.py` module and others get an actual name, e.g. `receivers.py`, nonetheless many seem to be APIs. 
- A clear separation between whats a model and what's an API, and even maybe splitting some of those API classes.
- I'd suggest to have an `/api` folder and then tiny classes inside since it seems this module does not need an actual service layer but it can be okay with just an api to bridge services (in rdm-records) to the GH api.

File: setup.cfg
Comment (alejandromumo) [reply setup.cfg:85]: It was added for tests. We have some vocabularies that are required (e.g. resource_type)

My bad on the typo 😂  I suppose it was being installed anyway by a dependency of this module?

File: invenio_github/api.py
Comment (alejandromumo) [reply invenio_github/api.py:1]: I totally agree. 
However, I'd leave tackle it separately

File: invenio_github/config.py
Comment (alejandromumo) [reply invenio_github/config.py:74]: So I left them - even if they are failing to load - because we're working on the UI on top of this PR. So I made sure that the building worked (e.g. added webpack configs) and tested the UI endpoints (e.g. they return the data needed for the templates to work).

File: LICENSE
Comment (alejandromumo) [reply LICENSE:1]: To be honest, I thought copyright strings are usually like `©2023 <COPYRIGHTS_HOLDER>` and that's why I left the current year only. I suppose a range is only accepted

File: invenio_github/tasks.py
Comment (alejandromumo) [reply invenio_github/tasks.py:80]: Agree, however, this was already there and I didn't want to mess too much with it:

https://github.com/inveniosoftware/invenio-github/blob/master/invenio_github/tasks.py#L82

I think that we need to have a cleanup phase after this is merged ...

File: invenio_github/utils.py
Comment (alejandromumo) [reply invenio_github/utils.py:1]: It was missing some cleanup actually ... methods that were API specific were already moved somewhere else. Thanks for the heads up!

File: invenio_github/utils.py
Comment (alejandromumo) [reply invenio_github/utils.py:86]: Added a config 👍 

Why 30? I am not sure

File: invenio_github/config.py
Comment (alejandromumo) [reply invenio_github/config.py:68]: Removed it straight away

File: invenio_github/views/github.py
Comment (alejandromumo) [reply invenio_github/views/github.py:75]: ping @anikachurilova

File: LICENSE
Comment (ppanero) [reply LICENSE:1]: It's not very clear in the docs https://invenio.readthedocs.io/en/latest/community/contributing/maintainers-guide/copyright-policy.html. But the idea is not to change is but to track since when, and we only modify when we change the file in a non trivial way. In this case, it's a bit more tricky duet to the licensing change.

File: invenio_github/api.py
Comment (alejandromumo) [inline invenio_github/api.py:475]: This feels a bit out of place here. 

The problem was that to publish a record, we needed the user's identity. Therefore, we used a utility function from invenio-access to generate this identity given the user id. 

But, there is permission needed to publish a draft (`authenticated_user`) and it was added here.

Possible alternatives:

1 - move this to `invenio-rdm-records` as this is a specific RDM action (publish)
2 - move this to `invenio-access` as a new utility function (e.g. `get_authenticated_identity`)

Comment (alejandromumo) [reply invenio_github/api.py:475]: Created an [issue](https://github.com/inveniosoftware/invenio-github/issues/105)

File: invenio_github/api.py
Comment (alejandromumo) [reply invenio_github/api.py:1]: Created an [issue](https://github.com/inveniosoftware/invenio-github/issues/103)

File: invenio_github/errors.py
Comment (alejandromumo) [reply invenio_github/errors.py:35]: Here I was not sure whether to include specific info or not. E.g. some errors might be revealed to the user, thus exposing things like `user_id` or `repository_id`. I do agree that, if we are not using that information, it'd be better if we remove it.  

Since it was already like this, I just followed what we had and postpone that decision for later.

Comment (alejandromumo) [reply invenio_github/errors.py:35]: Created an [issue](https://github.com/inveniosoftware/invenio-github/issues/106)

**Review by zzacharo — APPROVED** [review]

**Pull request merged**
